### ISSUE TITLE
AttributeError: module 'keras.engine' has no attribute 'topology'

### ISSUE THREAD

**Issue by johndpope** [PacktPublishing/Machine-Learning-with-Core-ML#5]
```python
import warnings
with warnings.catch_warnings():warnings.filterwarnings("ignore", category=RuntimeWarning)

import coremltools


coreml_model = coremltools.converters.keras.convert(
'tinyyolo_voc2007_modelweights.h5',
input_names='image',
image_input_names='image',
output_names='output',
image_scale=1./255.)

coreml_model.author = 'Joshua Newnham'
coreml_model.license = 'BSD'
coreml_model.short_description = 'Keras port of YOLOTiny VOC2007 by Joseph Redmon and Ali Farhadi'
coreml_model.input_description['image'] = '416x416 RGB Image'
coreml_model.output_description['output'] = '13x13 Grid made up of: [confidence, cx, cy, w, h, 20 x classes] * 5 bounding boxes'

coreml_model.save('tinyyolo_voc2007.mlmodel')


```
(tensorflow) ➜  Notebooks python test.py
WARNING:root:Keras version 2.2.0 detected. Last version known to be fully compatible of Keras is 2.1.3 .
WARNING:root:TensorFlow version 1.8.0 detected. Last version known to be fully compatible is 1.5.0 .
Traceback (most recent call last):
  File "test.py", line 12, in <module>
    image_scale=1./255.)
  File "/Users/jpope/miniconda2/envs/tensorflow/lib/python3.6/site-packages/coremltools/converters/keras/_keras_converter.py", line 745, in convert
    custom_conversion_functions=custom_conversion_functions)
  File "/Users/jpope/miniconda2/envs/tensorflow/lib/python3.6/site-packages/coremltools/converters/keras/_keras_converter.py", line 527, in convertToSpec
    from . import _keras2_converter
  File "/Users/jpope/miniconda2/envs/tensorflow/lib/python3.6/site-packages/coremltools/converters/keras/_keras2_converter.py", line 64, in <module>
    _keras.engine.topology.InputLayer:_layers2.default_skip,
AttributeError: module 'keras.engine' has no attribute 'topology'

**Comment by johndpope** [comment]
➜  Notebooks pip install 'keras==1.2.2' --force-reinstall
Collecting keras==1.2.2
  Downloading https://files.pythonhosted.org/packages/75/3e/9926ce5c678b7a7978724a2ecf24857d89a415d152b8d3443e6d45c228b2/Keras-1.2.2.tar.gz (175kB)
    100% |████████████████████████████████| 184kB 2.2MB/s
Requirement already satisfied: theano in /Users/jpope/miniconda2/lib/python2.7/site-packages (from keras==1.2.2)
Requirement already satisfied: pyyaml in /Users/jpope/miniconda2/lib/python2.7/site-packages (from keras==1.2.2)
Requirement already satisfied: six in /Users/jpope/miniconda2/lib/python2.7/site-packages (from keras==1.2.2)
Requirement already satisfied: scipy>=0.14 in /Users/jpope/miniconda2/lib/python2.7/site-packages (from theano->keras==1.2.2)
Requirement already satisfied: numpy>=1.9.1 in /Users/jpope/miniconda2/lib/python2.7/site-packages (from theano->keras==1.2.2)
Building wheels for collected packages: keras
  Running setup.py bdist_wheel for keras ... done
  Stored in directory: /Users/jpope/Library/Caches/pip/wheels/55/07/cf/b32db0a8d243b2fd6759d5d7cb650aa20670b2b740209cbf7e
Successfully built keras
Installing collected packages: keras
  Found existing installation: Keras 2.1.6
    Uninstalling Keras-2.1.6:
      Successfully uninstalled Keras-2.1.6
Successfully installed keras-1.2.2
You are using pip version 9.0.1, however version 18.0 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
➜  Notebooks python test.py
Using TensorFlow backend.
Traceback (most recent call last):
  File "test.py", line 12, in <module>
    image_scale=1./255.)
  File "/Users/jpope/miniconda2/lib/python2.7/site-packages/coremltools/converters/keras/_keras_converter.py", line 271, in convert
    raise RuntimeError('keras not found or unsupported version or backend found. keras conversion API is disabled.')
RuntimeError: keras not found or unsupported version or backend found. keras conversion API is disabled.

**Comment by johndpope** [comment]
need to do a pip freeze to spit out the keras / coreml pypi versions.

**Comment by johndpope** [comment]
pip install -U coremltools
pip install -U keras

pip freeze
```


alabaster==0.7.10
appnope==0.1.0
asn1crypto==0.22.0
astroid==1.5.3
Babel==2.4.0
backports-abc==0.5
backports.functools-lru-cache==1.4
backports.shutil-get-terminal-size==1.0.0
backports.weakref==1.0rc1
beautifulsoup4==4.6.0
bleach==1.5.0
certifi==2017.4.17
cffi==1.10.0
chardet==3.0.4
click==6.7
conda==4.3.30
configparser==3.5.0
coremltools==0.8
coverage==3.7.1
cryptography==1.8.1
cycler==0.10.0
decorator==4.0.11
docutils==0.13.1
ecdsa==0.13
entrypoints==0.2.3
enum34==1.1.6
funcsigs==1.0.2
functools32==3.2.3.post2
futures==3.2.0
goslate==1.5.1
h5py==2.8.0
html5lib==0.9999999
idna==2.5
imagesize==0.7.1
ipaddress==1.0.18
ipykernel==4.6.1
ipython==5.4.1
ipython-genutils==0.2.0
isort==4.2.15
jedi==0.10.2
Jinja2==2.9.6
jsonschema==2.6.0
jupyter-client==5.1.0
jupyter-core==4.3.0
Keras==2.2.2
Keras-Applications==1.0.4
Keras-Preprocessing==1.0.2
lazy-object-proxy==1.3.1
logilab-common==1.4.0
Markdown==2.6.8
MarkupSafe==1.0
matplotlib==2.0.2
mccabe==0.6.1
mistune==0.7.4
mock==2.0.0
mxnet==0.10.0
nbconvert==5.2.1
nbformat==4.3.0
networkx==1.11
nltk==3.3
numpy==1.15.0
numpydoc==0.7.0
olefile==0.44
opencv-python==3.2.0.7
packaging==16.8
pandocfilters==1.4.1
pathlib==1.0.1
pathlib2==2.3.0
Pattern==2.6
pbr==3.1.1
pep8==1.7.0
pexpect==4.2.1
pickleshare==0.7.4
Pillow==4.1.1
prompt-toolkit==1.0.14
protobuf==3.6.1
psutil==5.2.2
ptyprocess==0.5.2
pycosat==0.6.2
pycparser==2.17
PyDictionary==1.5.2
pyflakes==1.5.0
Pygments==2.2.0
pylint==1.7.2
pyOpenSSL==17.0.0
pyparsing==2.1.4
pypng==0.0.18
python-dateutil==2.6.0
pytz==2017.2
PyWavelets==0.5.2
PyYAML==3.13
pyzmq==16.0.2
QtAwesome==0.4.4
qtconsole==4.3.0
QtPy==1.2.1
requests==2.18.1
rope==0.10.5
scandir==1.5
scikit-image==0.13.0
scikit-learn==0.18.1
scipy==1.1.0
simplegeneric==0.8.1
simplejson==3.11.1
singledispatch==3.4.0.3
six==1.11.0
sklearn==0.0
snowballstemmer==1.2.1
Sphinx==1.6.2
sphinxcontrib-websupport==1.0.1
spyder==3.1.4
subprocess32==3.2.7
tensorflow==1.2.1
testpath==0.3.1
Theano==0.9.0
tornado==4.5.1
traitlets==4.3.2
typing==3.6.1
unicodecsv==0.14.1
urllib3==1.21.1
wcwidth==0.1.7
webencodings==0.5.1
Werkzeug==0.12.2
wrapt==1.10.10


```
Traceback (most recent call last):
  File "test.py", line 7, in <module>
    coreml_model = coremltools.converters.keras.convert(
AttributeError: 'module' object has no attribute 'convert'

**Comment by johndpope** [comment]
without requirements file - this is a battle to get python working. 



(coreml) ➜  Notebooks python test.py
Traceback (most recent call last):
  File "test.py", line 4, in <module>
    import coremltools
  File "/Users/jpope/miniconda2/envs/coreml/lib/python2.7/site-packages/coremltools/__init__.py", line 29, in <module>
    import converters
  File "/Users/jpope/miniconda2/envs/coreml/lib/python2.7/site-packages/coremltools/converters/__init__.py", line 10, in <module>
    import keras
  File "/Users/jpope/miniconda2/envs/coreml/lib/python2.7/site-packages/coremltools/converters/keras/__init__.py", line 15, in <module>
    from ._keras_converter import convert
  File "/Users/jpope/miniconda2/envs/coreml/lib/python2.7/site-packages/coremltools/converters/keras/_keras_converter.py", line 17, in <module>
    import _topology
  File "/Users/jpope/miniconda2/envs/coreml/lib/python2.7/site-packages/coremltools/converters/keras/_topology.py", line 24, in <module>
    _keras.layers.advanced_activations.ParametricSoftplus,
AttributeError: 'module' object has no attribute 'ParametricSoftplus'



@sagarsawant69 - I beseech you to use miniconda and export the environment to yml file - this will help python code run consistently across environments. Avoid using pip where possible - this will peg python code to version2 or 3 and avoid needless wasting of time trying to get out of date pip packages working. 

So - I have a resolution using miniconda to peg the coreml / keras / tensorflow version to older (deprecated versions) 

Iinstall mini conda to help sanitize python environments. 
 (otherwise cherry pick specific versions of pip in the link below)
https://conda.io/miniconda.html

https://gist.github.com/johndpope/650d61c5e550dbc2e09dbbeabfb0d95d

**Comment by johndpope** [comment]
so while the code above can successfully spit out the model - it is not working now with xcode Version 9.4.1 (9F2000). 

<img width="1533" alt="screen shot 2018-08-16 at 1 57 19 pm" src="https://user-images.githubusercontent.com/289994/44225765-5599b500-a15c-11e8-8b18-cacd43093f79.png">

**Comment by joshnewnham** [comment]
Thanks again for highlighting the issues - a slight oversight from myself. My intention was to avoid unnecessary work for the reader to have them concentrate on the concepts, CoreML and applying of it and it's for this reason that I opted to use https://notebooks.azure.com/ (which works as per the instructions within the book). But appreciate many will want to delve into the details and build locally (and to mitigate varying version of Azure notebooks); I have added an environment file (as you had suggested - thanks) within the Notebooks directories for Chapter05 and Chapter06 for referenced notebooks (along with comments within the relevant notebooks themselves). 
Thanks again for highlighting the issue - much appreciated and hope you're enjoying the book (despite the additional work you're having to do).
